File: Meadowlark-js/packages/meadowlark-core/src/middleware/ParsePathMiddleware.ts
// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import { writeDebugStatusToLog, writeRequestToLog } from '../Logger';
import { isDocumentUuidWellFormed } from '../model/DocumentIdentity';
import type { PathComponents } from '../model/PathComponents';
import type { DocumentUuid } from '../model/IdTypes';
import type { MiddlewareModel } from './MiddlewareModel';

const moduleName = 'core.middleware.ParsePathMiddleware';

export function pathComponentsFrom(path: string): PathComponents | null {
  // Matches all of the following sample expressions:
  // /v3.3b/ed-fi/Sections
  // /v3.3b/ed-fi/Sections/
  // /v3.3b/ed-fi/Sections/idValue
  const pathExpression = /\/(?<version>[^/]+)\/(?<namespace>[^/]+)\/(?<resource>[^/]+)(\/|$)((?<documentUuid>[^/]*$))?/gm;
  const match = pathExpression.exec(path);

  if (match?.groups == null) {
    return null;
  }

  const { documentUuid } = match.groups ?? null;

  return {
    version: match.groups.version,
    namespace: match.groups.namespace,
    resourceName: match.groups.resource,
    documentUuid: documentUuid as DocumentUuid,
  };
}

/**
 * Handles path parsing
 */
export async function parsePath({ frontendRequest, frontendResponse }: MiddlewareModel): Promise<MiddlewareModel> {
  // if there is a response already posted, we are done
  if (frontendResponse != null) return { frontendRequest, frontendResponse };
  writeRequestToLog(moduleName, frontendRequest, 'parsePath');

  const pathComponents: PathComponents | null = pathComponentsFrom(frontendRequest.path);

  if (pathComponents === null) {
    writeDebugStatusToLog(moduleName, frontendRequest, 'parsePath', 404);
    return { frontendRequest, frontendResponse: { statusCode: 404 } };
  }

  const { documentUuid } = pathComponents;
  if (documentUuid != null && !isDocumentUuidWellFormed(documentUuid)) {
    writeDebugStatusToLog(moduleName, frontendRequest, 'parsePath', 404, `Malformed resource id ${documentUuid}`);
    return { frontendRequest, frontendResponse: { statusCode: 404 } };
  }

  frontendRequest.middleware.pathComponents = pathComponents;
  return { frontendRequest, frontendResponse: null };
}

File: Meadowlark.net/Meadowlark.Net.Core/Middleware/ParsePathMiddleware.cs
using System.Text.RegularExpressions;
using Meadowlark.Net.Core.Model;

namespace Meadowlark.Net.Core.Middleware;

public static class ParsePathMiddleware
{
  private static string Decapitalize(string str)
  {
    if (str.Length == 0) return str;
    if (str.Length == 1) return str.ToLower();
    return str[0..1].ToLower() + str[1..];
  }

  private static PathComponents? PathComponentsFrom(string path)
  {
    // Matches all of the following sample expressions:
    // /ed-fi/Sections
    // /ed-fi/Sections/
    // /ed-fi/Sections/idValue
    Regex pathExpression = new(@"\/(?<projectNamespace>[^/]+)\/(?<endpointName>[^/]+)(\/|$)((?<documentUuid>[^/]*$))?");

    Match? match = pathExpression.Match(path);

    if (match == null)
    {
      return null;
    }

    string? documentUuidValue = match.Groups["documentUuid"].Value;
    if (documentUuidValue == "") documentUuidValue = null;

    return new(
      ProjectNamespace: new(match.Groups["projectNamespace"].Value.ToLower()),
      EndpointName: new(Decapitalize(match.Groups["endpointName"].Value)),
      DocumentUuid: documentUuidValue == null ? null : new(documentUuidValue)
    );
  }

  /**
   * Check that this is a well formed UUID v4
   */
  private static bool IsDocumentUuidWellFormed(DocumentUuid documentUuid)
  {
    // Regex for a UUID v4 string
    Regex uuid4Regex = new(@"^[0-9a-f]{8}-[0-9a-f]{4}-[4][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$");

    return uuid4Regex.IsMatch(documentUuid.Value.ToLower());
  }

  /**
   * Handles path parsing
   */
  public static MiddlewareModel ParsePath(MiddlewareModel middlewareModel)
  {
    var (requestModel, frontendResponse) = middlewareModel;

    // if there is a response already posted, we are done
    if (frontendResponse != null) return middlewareModel;

    PathComponents? pathComponents = PathComponentsFrom(requestModel.FrontendRequest.Path);

    if (pathComponents == null)
    {
      return middlewareModel with { FrontendResponse = new(StatusCode: 404, Body: "") };
    }

    var updatedMiddlewareModel = middlewareModel with { RequestModel = requestModel with { PathComponents = pathComponents } };

    if (pathComponents.DocumentUuid != null && !IsDocumentUuidWellFormed(pathComponents.DocumentUuid.Value))
    {
      return updatedMiddlewareModel with { FrontendResponse = new(StatusCode: 404, Body: "") };
    }

    return updatedMiddlewareModel;
  }
}
